File: src/main/java/com/github/penguin418/todolist/config/TodoUserDetailService.java
package com.github.penguin418.todolist.config;

import com.github.penguin418.todolist.mapper.TodoUserMapper;
import com.github.penguin418.todolist.model.dto.TodoUserDto;
import com.github.penguin418.todolist.model.entity.TodoUserEntity;
import com.github.penguin418.todolist.model.request.WithdrawalRequest;
import com.github.penguin418.todolist.model.request.JoinRequest;
import com.github.penguin418.todolist.repository.TodoRepository;
import com.github.penguin418.todolist.repository.TodoUserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;

@Component
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class TodoUserDetailService implements UserDetailsService {
    private final PasswordEncoder passwordEncoder;
    private final TodoUserRepository todoUserRepository;
    private final TodoRepository todoRepository;
    private final TodoUserMapper todoUserMapper;

    @Override
    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
        return todoUserRepository.findByUsername(username)
                .map(todoUserMapper::toAuthentication)
                .orElseThrow(() -> new UsernameNotFoundException(username));
    }

    @Transactional
    public TodoUserDto createAccount(JoinRequest joinRequest) {
        if (todoUserRepository.findByUsername(joinRequest.getUsername()).isPresent()) {
            throw new IllegalArgumentException("Duplicated username");
        }

        TodoUserEntity todoUser = new TodoUserEntity(
                null,
                joinRequest.getUsername(),
                passwordEncoder.encode(joinRequest.getPassword()),
                joinRequest.getNickname()
        );
        todoUserRepository.save(todoUser);
        return todoUserMapper.toDto(todoUser);
    }

    public TodoUserDto getTodoUserDto(TodoUser todoUser) {
        return todoUserMapper.toDto(todoUser);
    }

    @Transactional
    public void deleteAccount(TodoUser todoUser, WithdrawalRequest withdrawalRequest) {
        if (!passwordEncoder.matches(withdrawalRequest.getPassword(), todoUser.getPassword())) {
            throw new IllegalArgumentException("Password not matched");
        }
        todoRepository.deleteByTodoUserId(todoUser.getUserId());
        todoUserRepository.deleteById(todoUser.getUserId());
    }
}

File: src/main/java/com/github/penguin418/todolist/config/TodoUserDetailService.kt
package com.github.penguin418.todolist.config

import com.github.penguin418.todolist.mapper.TodoUserMapper
import com.github.penguin418.todolist.model.dto.TodoUserDto
import com.github.penguin418.todolist.model.entity.TodoUserEntity
import com.github.penguin418.todolist.model.request.JoinRequest
import com.github.penguin418.todolist.model.request.WithdrawalRequest
import com.github.penguin418.todolist.repository.TodoRepository
import com.github.penguin418.todolist.repository.TodoUserRepository
import org.springframework.security.authentication.BadCredentialsException
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.security.core.userdetails.UsernameNotFoundException
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional(readOnly = true)
open class TodoUserDetailService(
        val passwordEncoder: PasswordEncoder,
        val todoUserRepository: TodoUserRepository,
        val todoRepository: TodoRepository,
        val todoUserMapper: TodoUserMapper) : UserDetailsService {
    @Throws(UsernameNotFoundException::class)
    override fun loadUserByUsername(username: String): UserDetails {
        return todoUserRepository.findByUsername(username)
                .map { entity: TodoUserEntity? -> todoUserMapper.toAuthentication(entity) }
                .orElseThrow { UsernameNotFoundException(username) }
    }

    @Transactional
    open fun createAccount(joinRequest: JoinRequest): TodoUserDto {
        check(!todoUserRepository.findByUsername(joinRequest.username!!).isPresent) { "Duplicated username" }
        val todoUser = TodoUserEntity(
                null,
                joinRequest.username,
                passwordEncoder.encode(joinRequest.password),
                joinRequest.nickname
        )
        todoUserRepository.save(todoUser)
        return todoUserMapper.toDto(todoUser)
    }

    fun getTodoUserDto(todoUser: TodoUser?): TodoUserDto {
        return todoUserMapper.toDto(todoUser)
    }

    @Transactional
    open fun deleteAccount(todoUser: TodoUser, withdrawalRequest: WithdrawalRequest) {
        if (!passwordEncoder.matches(withdrawalRequest.password, todoUser.getPassword())) {
            throw BadCredentialsException("Password not matched")
        }
        todoRepository.deleteByTodoUserId(todoUser.userId)
        todoUserRepository.deleteById(todoUser.userId)
    }
}
